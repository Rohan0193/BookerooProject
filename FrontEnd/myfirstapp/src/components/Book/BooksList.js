import React, {useEffect, useState} from 'react'
import {bookMicroServiceApi} from '../../Utilities/hosts'


//a functional component that fetches all books from back-end api and displays a search bar to search relevant books
const BooksList = () => {

    //create a hook to catch and display errors
    const [error, setError] = useState(null);
    //create a hook to check if books are loaded
    const [isLoaded, setIsLoaded] = useState(false);
    //create a hook to save state of book
    const [bookList, setItems] = useState([]);
    //hook to store state of search keyword
    const [query, setquery] = useState("");
    //different paramters to filter search results
    const [searchParam] = useState(["title", "isbn","genre", "authorName"]);

    //Effect hook to fetch books from the API
    useEffect( () => {
        fetch(bookMicroServiceApi + "/books")
        .then(response => response.json())
        .then(
            (result) => {
                setIsLoaded(true);
                setItems(result);
                console.log(result);
            },
            (error) => {
                setIsLoaded(true);
                setError(error);
                }
            )
        },[])

   //function to filter items based on given paramters
   function search(bookItems){
       return bookItems.filter((item) => {
           return searchParam.some((newBookList) => {
               return (
                   item[newBookList]
                        .toString()
                        .toLowerCase()
                        .indexOf(query.toLocaleLowerCase()) > -1      
               );
           });
       });
   }
   //handle fetch errors, loading and displaying of information of the page
   if(error){
       return <div>Error : {error.message}</div>;
   } else if(!isLoaded){
       return <div>Loading...</div>;
   }else{
       return (
            <div className="wrapper">
                <div> 
                <input 
                    type="text" 
                    className="search-bar" 
                    placeholder="Filter items"
                    value = {query}
                    onChange={(e) => setquery(e.target.value)}
                    />
                </div>
                <ul className="card-grid">
                    {/*loop through the list array created by the hook state to display info*/}
                    {search(bookList).map((books) => (
                        <li>
                            <a href={`book/${books.isbn}`}>
                                <article className="card cards" key={books.callingCodes}>
                                    <div className="card-image bookImg">
                                        <img src={`http://covers.openlibrary.org/b/isbn/${books.isbn}-M.jpg`} alt={books.name} />
                                    </div>
                                    <div className="card-content">
                                        <h2 className="card-name">{books.name}</h2>
                                        <ol className="card-list">
                                            <li>
                                                Book ISBN:{" "}
                                                <span>{books.isbn}</span>
                                            </li>
                                            <li>
                                                Book Title: <span>{books.title}</span>
                                            </li>
                                            <li>
                                                Book Author: <span>{books.authorName}</span>
                                            </li>
                                        </ol>
                                    </div>
                                </article>
                            </a>
                        </li>
                    ))}
                </ul>
            </div>
       );
   }
}

export default BooksList;