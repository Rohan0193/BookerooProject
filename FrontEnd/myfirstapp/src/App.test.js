import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer'

// setup file
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import BooksList from './components/BookDisplay/BooksList';

configure({ adapter: new Adapter() });

//snapshot testing
//bookslist
describe('Snapshot Testing -- BooksList', () => {
  it('Test to see if component works as expected',() => {
    const component = renderer.create(<BooksList />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})


// describe('Test useEffect() function in books list', () => {
//   it('Gets the right amount of entries in the database', () =>{
//      shallow( 
//             <BooksList />
//             ).dive()
//             expect(wrapper.find('p').text()).toContain("There are 1 in the database!")
//   })
// })
