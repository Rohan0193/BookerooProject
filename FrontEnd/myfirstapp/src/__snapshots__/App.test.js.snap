// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing -- BooksList Test to see if component works as expected 1`] = `
<div>
  Loading...
</div>
`;
